File: app/src/main/java/de/tum/in/tumcampus/models/CafeteriaMenu.java
package de.tum.in.tumcampus.models;

import android.content.Context;
import android.text.SpannableString;
import android.text.style.ImageSpan;

import java.util.Date;

import de.tum.in.tumcampus.R;
import de.tum.in.tumcampus.auxiliary.Utils;

/**
 * CafeteriaMenu object
 */
public class CafeteriaMenu {

	/**
	 * Cafeteria ID
	 */
	public int cafeteriaId;

	/**
	 * Menu date
	 */
	public Date date;

	/**
	 * CafeteriaMenu Id (empty for addendum)
	 */
	public int id;

	/**
	 * Menu name
	 */
	public String name;

	/**
	 * Long type, e.g. Tagesgericht 1
	 */
	public String typeLong;

	/**
	 * Type ID
	 */
	public int typeNr;

	/**
	 * Short type, e.g. tg
	 */
	public String typeShort;

	// public String prize;

	/**
	 * New CafeteriaMenu
	 * 
	 * <pre>
	 * @param id CafeteriaMenu Id (empty for addendum)
	 * @param cafeteriaId Cafeteria ID
	 * @param date Menu date
	 * @param typeShort Short type, e.g. tg 
	 * @param typeLong Long type, e.g. Tagesgericht 1
	 * @param typeNr Type ID
	 * @param name Menu name
	 * </pre>
	 */
	public CafeteriaMenu(int id, int cafeteriaId, Date date, String typeShort,
			String typeLong, int typeNr, String name) {

		this.id = id;
		this.cafeteriaId = cafeteriaId;
		this.date = date;
		this.typeShort = typeShort;
		this.typeLong = typeLong;
		this.typeNr = typeNr;
		this.name = name;
		// this.prize = prize;
	}

    @Override
	public String toString() {
		return "id=" + this.id + " cafeteriaId=" + this.cafeteriaId + " date="
				+ Utils.getDateString(this.date) + " typeShort="
				+ this.typeShort + " typeLong=" + this.typeLong + " typeNr="
				+ this.typeNr + " name=" + this.name;
	}
}

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/cafeteria/model/CafeteriaMenu.kt
package de.tum.`in`.tumcampusapp.component.ui.cafeteria.model

import android.content.Context
import androidx.room.Entity
import androidx.room.PrimaryKey
import androidx.room.RoomWarnings
import com.google.gson.annotations.SerializedName
import de.tum.`in`.tumcampusapp.component.ui.cafeteria.model.deserialization.DishPrices
import org.joda.time.DateTime
import java.util.regex.Pattern

/**
 * CafeteriaMenu
 *
 * @param menuId CafeteriaMenu Id (empty for addendum)
 * @param cafeteriaId primary key of the associated cafeteria
 * @param slug cafeteria string identification slug e.g.: "mensa-garching"
 * @param date date on which the dish is served
 * @param dishType
 * @param name
 * @param labels Labels describing the dish. Typically these are allergens.
 * @param calendarWeek calendar week for the current year in which the dish is served
 */
@Entity
@SuppressWarnings(RoomWarnings.DEFAULT_CONSTRUCTOR)
data class CafeteriaMenu(
        @PrimaryKey(autoGenerate = true)
        @SerializedName("menuId")
        var menuId: Int = 0,
        @SerializedName("cafeteriaId")
        var cafeteriaId: Int = 0,
        @SerializedName("cafeteriaSlug")
        var slug: String = "",
        @SerializedName("date")
        var date: DateTime? = null,
        @SerializedName("dishType")
        var dishType: String = "",
        @SerializedName("name")
        var name: String = "",
        @SerializedName("labels")
        var labels: String = "",
        @SerializedName("calendarWeek")
        var calendarWeek: Int = -1,
        @SerializedName("dishPrices")
        var dishPrices: DishPrices
) {

    val tag: String
        get() = "${name}__$cafeteriaId"

    // TODO pricing, dishType enum, label enum?, mensa enum?
    val menuType: MenuType
        get() {
            return when (dishType) {
                "tg" -> MenuType.DAILY_SPECIAL
                "ae" -> MenuType.DISCOUNTED_COURSE
                "akt" -> MenuType.SPECIALS
                "bio" -> MenuType.BIO
                else -> MenuType.SIDE_DISH
            }
        }

    val notificationTitle: String
        get() {
            return REMOVE_DISH_ENUMERATION_PATTERN
                    .matcher(dishType)
                    .replaceAll("")
                    .trim()
        }

    fun getNotificationText(context: Context): String {
        val lines = getNotificationLines(context)
        return if (dishType == MenuType.SPECIALS.toString()) {
            lines.joinToString(", ")
        } else {
            lines.first()
        }
    }

    private fun getNotificationLines(context: Context): List<String> {
        return if (dishType == MenuType.SPECIALS.toString()) {
            // Returns a list of all specials
            name
                    .split("\n")
                    .map { it.trim() }
        } else {
            // Returns a list containing the dish name and the price
            val priceText = getPriceText(context)
            listOfNotNull(name, priceText)
        }
    }

    fun getPriceText(context: Context): String {
        val rolePrice = dishPrices.getRolePrice(context)
        return if (rolePrice.basePrice != 0.0) String.format("%.2f€ + %.2f€", rolePrice.basePrice, rolePrice.pricePerUnit)
        else String.format("%.2f€", rolePrice.pricePerUnit)
    }

    companion object {
        private val REMOVE_DISH_ENUMERATION_PATTERN: Pattern = Pattern.compile("[0-9]")
    }
}
